library/core/src/macros/mod.rs

```diff
@@ -45,24 +46,38 @@ macro_rules! assert_eq {
                     // The reborrows below are intentional. Without them, the stack slot for the
                     // borrow is initialized even before the values are compared, leading to a
                     // noticeable slow down.
-                    $crate::panic!(r#"assertion failed: `(left == right)`
-  left: `{:?}`,
- right: `{:?}`"#, &*left_val, &*right_val)
+                    let left_dbg: &dyn $crate::fmt::Debug = &&*left_val;
+                    let right_dbg: &dyn $crate::fmt::Debug = &&*right_val;
+                    $crate::panicking::panic_cmp(
+                        "assert_eq",
+                        "==",
+                        stringify!($left),
+                        stringify!($right),
+                        left_dbg,
+                        right_dbg,
+                    );
                 }
             }
         }
     });
-    ($left:expr, $right:expr, $($arg:tt)+) => ({
-        match (&($left), &($right)) {
+    ($left:expr, $right:expr, $($arg:tt)*) => ({
```

Comment:
Why was this changed from + to *? Does that make a difference?

Reply:
It was intentionally changed to + in commit 65435e14febd9d8aa1d40a19cd1adb427263c48e. But in that time, there was no `$(,)?` in the first matcher.